locator/abstract_replication_strategy.cc

```diff
@@ -308,10 +308,22 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
 
 future<mutable_effective_replication_map_ptr> calculate_effective_replication_map(abstract_replication_strategy::ptr_type rs, token_metadata_ptr tmptr) {
     replication_map replication_map;
-
-    for (const auto &t : tmptr->sorted_tokens()) {
-        auto eps = co_await rs->calculate_natural_endpoints(t, *tmptr);
-        replication_map.emplace(t, eps.get_vector());
+    const auto& sorted_tokens = tmptr->sorted_tokens();
+
+    if (!sorted_tokens.empty()) {
```

Comment:
Under what conditions can this be empty here? And how do callers handle this?
Not saying this check is wrong, I'm just not that familiar with this code and would like to understand.

Reply:
I don't understand it either, except that I know we should work to make it impossible.

Reply:
It can definitely happen when starting in unit tests as seen here: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1780/artifact/testlog/x86_64/debug/

Reply:
I can add this to help understand how we get here:
```diff
diff --git a/locator/abstract_replication_strategy.cc b/locator/abstract_replication_strategy.cc
index fac202928d..198d26fa92 100644
--- a/locator/abstract_replication_strategy.cc
+++ b/locator/abstract_replication_strategy.cc
@@ -324,6 +324,8 @@ future<mutable_effective_replication_map_ptr> calculate_effective_replication_ma
                 co_await coroutine::maybe_yield();
             }
         }
+    } else {
+        rslogger.warn("calculate_effective_replication_map: token metadata is empty, at {}", current_backtrace());
     }
 
     auto rf = rs->get_replication_factor(*tmptr);
```

Reply:
FWIW, I've seen this case in the past, IIRC during early bootstrap state, before assigning tokens to the node.

Reply:
ok, so the following patch:
```diff
iff --git a/locator/abstract_replication_strategy.cc b/locator/abstract_replication_strategy.cc
index fac202928d..b64fdccd25 100644
--- a/locator/abstract_replication_strategy.cc
+++ b/locator/abstract_replication_strategy.cc
@@ -380,6 +380,10 @@ future<effective_replication_map_ptr> effective_replication_map_factory::create_
         co_return erm;
     }
 
+    if (tmptr->sorted_tokens().empty()) {
+        rslogger.warn("create_effective_replication_map: token metadata is empty, at {}", current_backtrace());
+    }
+
     // try to find a reference erm on shard 0
     // TODO:
     // - use hash of key to distribute the load
diff --git a/replica/database.cc b/replica/database.cc
index 8ecc411a99..448e0ed475 100644
--- a/replica/database.cc
+++ b/replica/database.cc
@@ -1354,6 +1354,7 @@ database::create_keyspace(const lw_shared_ptr<keyspace_metadata>& ksm, locator::
         co_return;
     }
 
+    dblog.info("Creating keyspace '{}': system={} bootstrap={}", ksm->name(), system, is_bootstrap);
     co_await create_in_memory_keyspace(ksm, erm_factory, system);
     auto& ks = _keyspaces.at(ksm->name());
     auto& datadir = ks.datadir();
```

shows that it happens when creating system_schema during bootstrap, for example in batchlog_manager_test:
```
INFO  2022-08-11 12:11:37,978 [shard 0] database - Creating keyspace 'system_schema': system=true bootstrap=true
WARN  2022-08-11 12:11:37,979 [shard 0] replication_strategy - create_effective_replication_map: token metadata is empty, at 0x...
```

and in alternator test (that starts the scylla binary):
```
INFO  2022-08-11 12:13:28,519 [shard 0] database - Creating keyspace 'system': system=true bootstrap=true
WARN  2022-08-11 12:13:28,519 [shard 0] replication_strategy - create_effective_replication_map: token metadata is empty, at 0x...
```

Reply:
It looks like we can pass the `is_bootstrap` parameter to `create_keyspace` down we can allow empty tokens only with it set to true, otherwise call on_internal_error.

Reply:
We can do that as a follow up as this series preserves the current behavior.